feat(color): add color component

jcfranco opened the pull request:
Issue #64

This adds the following components:
  - `calcite-color`
  - `calcite-color-hex-input`
  - `calcite-color-swatch`

### API

```ts
interface CalciteColor {
  // intl props omitted
  scale: "s" | "m" | "l";
  storageId: string;
  theme: "dark" | "light";
  value: ColorValue; 

  calciteColorChange: EventEmitter<void>;

  setFocus(): Promise<void>;
}

// CSS color | 
// { r: number, g: number: b: number} | 
// { h: number, s: number: v: number} |  
// { h: number, s: number: l: number};
type ColorValue = string | object; 

interface CalciteColorHexInput {
  hexLabel: string; // text used for aria-label
  scale: "s" | "m" | "l";
  theme: "dark" | "light";
  value: string;  // hex value

  onCalciteColorHexInputChange();

  setFocus(): Promise<void>;
}

interface CalciteColorSwatch {
  active: boolean;
  color: string;  // CSS color values
  scale: "s" | "m" | "l";
  theme: "dark" | "light";
}
```

### Pending

- [x] Confirm naming convention of label props (`intlProp`?).
- [x] Update canvas rendering code for HD displays (looks a bit blurry in those cases)
- [x] Update tests
- [x] a11y audit
- [x] Add stories

### Notes

* The format of value will be preserved as the user makes updates (e.g., if user specifies value as `rgb()`, it will remain `rgb()` after making updates).
* This adds https://www.npmjs.com/package/color as a dependency for working for color values.
* This version does not handle alpha (will be handled in separate issue)
* This also adds a new helper for testing component focus.
* The change event has no detail object for the following reasons:
  * input value format is preserved (assumes input value is the desired output one)
  * color can change frequently and adding all of the possible color formats to the event seems unnecessary
  * _secret_  :shipit: `color` prop is available for advanced color use cases
* Hex values are normalized to be lowercase
* Has groundwork for handling/preserving alpha values
* Simplifies `tsconfig.json`.

### Questions

* What should the default color be? It's currently set to `#007AC2`.
* Similar to `calcite-date`, should this be renamed to `calcite-color`?

jcfranco commented:
@bstifle I think this is ready for your assessment now. 😄 Any remaining things I'm working on shouldn't impact the HTML structure or styles. Let me know if I need to refactor anything to make styling changes easier.

bstifle commented:
@jcfranco 

hey man! awesome work so far. Functionality wise, seems to work as expected!

here's my comp:
![image](https://user-images.githubusercontent.com/14083530/83697415-cd480680-a5b3-11ea-9e98-d763a4f00245.png)




here are the things i'm seeing for fixes:

![image](https://user-images.githubusercontent.com/14083530/83697044-cb317800-a5b2-11ea-8352-0e4759f2198d.png)
**1.** Let's remove the negative space between the outline and the fill in the color shape. please make the outline border-2 (always)


![image](https://user-images.githubusercontent.com/14083530/83697279-6f1b2380-a5b3-11ea-8875-98ad7d8a731c.png)
**2.** Please make these ui-text-3, and weight medium, 14px. 12px font-size on small.


![image](https://user-images.githubusercontent.com/14083530/83697223-527eeb80-a5b3-11ea-8d0f-fc7349b224f0.png)
**3.** Please make selected version ui-text-1, medium, 14px font-size, with 3px ui-blue underline see comp. 12px font-size on small.


![image](https://user-images.githubusercontent.com/14083530/83697436-d8029b80-a5b3-11ea-93c2-93e03a1d3971.png)
**4.** Please remove all of these arrows in the inputs. really weird UX


![image](https://user-images.githubusercontent.com/14083530/83697482-f8caf100-a5b3-11ea-9d16-6a1f08d0f419.png)
**5.** Please match hover states on these interactions, i can provide specs if you need


![image](https://user-images.githubusercontent.com/14083530/83697513-0da78480-a5b4-11ea-9b7b-94e7220e1a2c.png)
**6.** Please match idle states of the +/- buttons foreground-1. foreground-2 on hover. foreground-3 on press. Please make hit state smaller, closer to 24x24px. Remove gutter between the elements please. 


![image](https://user-images.githubusercontent.com/14083530/83697568-36c81500-a5b4-11ea-9abb-f8b2baacfd11.png)
**7.** Please make input borders use 1px thickness and #949494 color


![image](https://user-images.githubusercontent.com/14083530/83697664-7131b200-a5b4-11ea-8099-03b97cfa74cf.png)
**8.** Please make divider lines use color border-3


![image](https://user-images.githubusercontent.com/14083530/83697751-ad651280-a5b4-11ea-940e-161b14b5eb00.png)
![image](https://user-images.githubusercontent.com/14083530/83697810-c8378700-a5b4-11ea-86c2-425733bd816c.png)
**9.** Focus states have a rounded radius, please remove.


![image](https://user-images.githubusercontent.com/14083530/83697849-d84f6680-a5b4-11ea-9912-99a18f7eb202.png)
![image](https://user-images.githubusercontent.com/14083530/83698001-2a908780-a5b5-11ea-9acd-3a34a69d3aa0.png)
**10.** Let's remove the gray outer border, add shadow-2 on idle, shadow-2-hover, shadow-2-press. 


![image](https://user-images.githubusercontent.com/14083530/83698036-472cbf80-a5b5-11ea-8815-9dd5d34f5926.png)
**11.** Please make both handles 20x20px with 3px white border thickness on Large and Medium. 16x16 on size Small with 3px white border.


![image](https://user-images.githubusercontent.com/14083530/83698200-b0143780-a5b5-11ea-93ba-b305fa842e57.png)
**12.** make color band 12 px high on Large and Medium. 8px on size Small.


![image](https://user-images.githubusercontent.com/14083530/83698411-3892d800-a5b6-11ea-9076-859e99bd00d5.png)
**13.** please make #, R, G, B ui-text-3, demibold, and 14px. 12px font-size on small.


![image](https://user-images.githubusercontent.com/14083530/83698642-cf5f9480-a5b6-11ea-9035-0dc6075ef937.png)
**14.** This one is kinda scary :)


![image](https://user-images.githubusercontent.com/14083530/83698683-ee5e2680-a5b6-11ea-892e-35d73a9b4973.png) 
![image](https://user-images.githubusercontent.com/14083530/83698703-f61dcb00-a5b6-11ea-8c93-5be04516a38c.png)
**15.** There are 2 separate focus states on this. the thin line one should be removed


![image](https://user-images.githubusercontent.com/14083530/83698801-2f563b00-a5b7-11ea-86a0-74e7cf623572.png)
**16.** Please keep dark theme handles white with shadow styling on #10


![image](https://user-images.githubusercontent.com/14083530/83699063-eeaaf180-a5b7-11ea-8f37-b74062348b79.png)
**17.** Doesn't look like inputs are using avenir-next

jcfranco commented:
1. 💥
2. Can you confirm if _medium_ === `font-weight: 500`? If so, 💥
3. In progress 🚧
4. 💥 FWIW, the up/down arrows are natively rendered by browsers for numeric inputs. It's useful if you're a mouse-forward user.
5. Let's sync-up on this one. I was trying to use `<calcite-button>`s internally
6. Gutters removed 💥, for the rest 👆
7. Border width fixed 💥 -> can you confirm whether #949494/#757575 suggestion is for the text or border?
8. 💥
9. As far as I can tell, we don't override default focus styling. The rounded focus got introduced in Chrome after https://developers.google.com/web/updates/2020/05/nic83 landed. 
10. Pending: This is going to take a little bit of work since I'm using `<canvas>` to render those elements.
11. 💥
12. Updated heights 💥, still need to tweak canvas rendering to fit hue slider thumb
13. 💥
14. I think the design for the small scale needs adjusting. Even after making adjustments, the inputs overflow (at least on Chrome).

<img width="208" alt="Screen Shot 2020-06-03 at 10 52 58 PM" src="https://user-images.githubusercontent.com/197440/83722533-ff785900-a5f1-11ea-8204-4f839b25c1c4.png">

15. This seems to be a `<calcite-button>` bug. I'll create an issue.
16. Using white thumb handles in dark theme 💥 Could you clarify the linked #10?
17. 💥
18. 💥

bstifle commented:
2. Can you confirm if _medium_ === `font-weight: 500`? If so, 💥

>  ✅

4. 💥 FWIW, the up/down arrows are natively rendered by browsers for numeric inputs. It's useful if you're a mouse-forward user. 

> **doesn't make sense to go up in 1 hex increments. not a use case**

5. Let's sync-up on this one. I was trying to use `<calcite-button>`s internally

> **ok**

7. Border width fixed 💥 -> can you confirm whether #949494/#757575 suggestion is for the text or border?

> **Border**

9. As far as I can tell, we don't override default focus styling. The rounded focus got introduced in Chrome after https://developers.google.com/web/updates/2020/05/nic83 landed. 

> **please work with adam on this. shouldn't pick up that browser style**

14. I think the design for the small scale needs adjusting. Even after making adjustments, the inputs overflow (at least on Chrome).

<img width="208" alt="Screen Shot 2020-06-03 at 10 52 58 PM" src="https://user-images.githubusercontent.com/197440/83722533-ff785900-a5f1-11ea-8204-4f839b25c1c4.png">

> **we probably cant use calcite inputs here unfortunately, we just need to get it to spec (smaller, narrower)** let's do height of 32px

16. Using white thumb handles in dark theme 💥 Could you clarify the linked #10?

> **Githubs markdown, no link associated, just the screenshot**

paulcpederson commented:
Nice! This is a ton of work, super cool. Really looking forward to using this one, we have a huge need for this.

> Similar to calcite-date, should this be renamed to `calcite-color`?

Yes, I think that makes sense. I like this because it makes it clear that the sub-component parts are used in this component. ie. it's more apparent that `calcite-color-swatch` is _part of_ `calcite-color`. 

---

After running locally, some thoughts:

1. Wherever we can, it would be good to leverage our own components in this color UI. Some ideas:
  - +/- buttons could be transparent, dark, small-scale `calcite-buttons`
  - the tabs could be `calcite-tabs`
  - inputs could be instances of `calcite-input` with prefix text
This will help us make these UI's consistent with everything else. Right now the line height of the inputs is incorrect, so the baselines aren't aligned, the hover state of the tab is missing, the +/- hover has no transition. Just little silly things like that we'd get for free by "drinking our own champagne", as it were. 
2. If we figured out a way to use `fill` instead of `color` for the swatches at the bottom, the would appear in high contrast
<img width="379" alt="Screen Shot 2020-07-22 at 8 57 06 AM" src="https://user-images.githubusercontent.com/1031758/88199458-be67f280-cbf9-11ea-8098-7db571b15eff.png">
3. Above image is firefox, looks like there may be layout issues? <br/>
4. The dark theme inputs aren't dark <img width="283" alt="Screen Shot 2020-07-22 at 8 29 03 AM" src="https://user-images.githubusercontent.com/1031758/88199540-d8093a00-cbf9-11ea-9303-232df26a88ff.png"> (This would go away if we used `calcite-input`
5. Getting a linter error about guid being unused in `calcite-color-picker.tsx`

---

This is all pretty minor stuff. Super impressed and thanks from the bottom of my heart 🎨 ✨ 🖍️

jcfranco commented:
@paulcpederson Thanks for reviewing! 🙇 

2–5 have been addressed. 🔧

Regarding 1, I agree and earlier iterations had `calcite-button` and `calcite-input` wired up, but there were some design inconsistencies. @bstifle WDYT?

bstifle commented:
@jcfranco nice work so far!

here are some changes.

The selected colors in the hex input and "saved colors" selection wasn't working out as hoped, so i simplified the borders on those.  here's all my updates

![image](https://user-images.githubusercontent.com/14083530/88234304-92666480-cc2d-11ea-8784-b7a78f51bbf9.png)

jcfranco commented:
Addressed all comments except for:

* opacity slider (will be tackled by https://github.com/Esri/calcite-components/issues/749)
* dogfooding components instead of duplicating styles/behaviors (I'll create a follow-up issue for this)

macandcheese commented:
Looking great @jcfranco - agree on naming it `calcite-color` as it matches `calcite-date`. Along those lines perhaps updating `calcite-hex-input` to share a `calcite-color-*` root would make sense, if it's meant to be available to use on its own.

Design wise, looking great! One thing, I know MVB has some instances where there is an input to select color w/out a full-blown color picker, just select from a list - maybe that specific design could be synced up with this (I feel like its a square swatch at the far right end vs. the "target circle" - which would be a nice re-use of the swatch item as well) @bstifle @mitc7862 @asangma

macandcheese commented on src/components/calcite-color-picker/calcite-color-picker.scss:
```diff
@@ -0,0 +1,278 @@
+  border-bottom: $color-mode-selected-border-height solid transparent;
+  transition: all 0.15s ease-in-out;
+
+  &:hover:not(.color-mode-selected) {
```
Does `calcite-tab` itself here not work?

jcfranco replied:
I'll do this [in a follow up issue](https://github.com/Esri/calcite-components/pull/592#issuecomment-662789967) (using `calcite-tab`, `calcite-button` and `calcite-input`).

macandcheese replied:
Rad! - yeah I just wired up the calcite-inputs to test with `scale=s` and `prefix-text` and it works with just a couple of CSS updates :) https://jmp.sh/PFzQ1KV

jcfranco replied:
That's awesome! Almost there! 🎉

jcfranco commented:
Renamed `calcite-color-picker` -> `calcite-color` and `calcite-hex-input` -> `calcite-color-hex-input`.

bstifle commented:
@jcfranco 

Hey man, after talking with adam and julio on this

here are some updates...

![image](https://user-images.githubusercontent.com/14083530/88343655-bd68bb00-ccf6-11ea-9e54-10216667dff9.png)

1. Use calcite-tabs
2. Use calcite input with the Small property, and prefixes for #, RGB, then suffix for opacity %
3. Put the selected hex color indicator absolute position on the right of the input, and make it 20x20px please
![image](https://user-images.githubusercontent.com/14083530/88343307-fb191400-ccf5-11ea-83fd-5907d5882462.png)

jcfranco commented:
Using `calcite-tabs` and `calcite-input` will be tackled separately. `calcite-input` still needs some changes before it can be dropped in place: 

* missing `maxLength` (and `minLength` although this one isn't used at the moment).
* input event canceling not working properly (text changes still go through)

cc @macandcheese 

>  make it 20x20px please

I'm using the swatch component for consistency and has the following sizes for scale (consistent w/ `calcite-icon`):

* s -> 16px
* m -> 24px
* l -> 32px

Should one of these sizes be changed to 20 or do I keep it at 24?

Here's a preview of the hex input layout changes & the swatch using the small scale:

![Screen Shot 2020-07-23 at 4 31 08 PM](https://user-images.githubusercontent.com/197440/88348341-040fe280-cd02-11ea-88df-fa377897ebe4.png)
![Screen Shot 2020-07-23 at 4 31 15 PM](https://user-images.githubusercontent.com/197440/88348345-04a87900-cd02-11ea-8e30-0902bdc00543.png)

Would you like me to proceed w/ 👆 for this PR or should I make these changes once the component is using `calcite-input`?

macandcheese commented:
> * missing `maxLength` (and `minLength` although this one isn't used at the moment).

I'll add these as explicit props re: https://github.com/Esri/calcite-components/issues/757, but they should at least be spread and work currently, I just tested locally and it seemed to work:

<img width="550" alt="Screen Shot 2020-07-23 at 5 00 56 PM" src="https://user-images.githubusercontent.com/4733155/88349917-4dfac780-cd06-11ea-9f2e-1959aa619e9e.png">

jcfranco commented:
>  they should at least be spread and work currently

Got it, but TS complains since the props don't exist on the component. 😞 

> Can you link to where you are doing this

Ignore this one for now. I referenced the wrong event and I think I'm doing something incorrect on my side.

jcfranco commented:
BTW, I've got a workaround for my input 'blockers'. 💪 I'll proceed with the latest design updates (using the small scale for the selected color preview for now).

jcfranco commented:
Almost done with the last changes. ✨ One thing came up with the small scale; the inputs won't fit properly:

<img width="196" alt="Screen Shot 2020-07-23 at 10 14 50 PM" src="https://user-images.githubusercontent.com/197440/88363179-f7a47d80-cd33-11ea-89d2-4100d1f273a0.png">

The component width when `scale=small` is 170px, and the minimum required for all of them to be legible is 230px. Another hurdle here is that the width used for the medium scale is 240px, so we can't just bump up the small scale width.

@bstifle @macandcheese @julio8a Please advise. 🤖

jcfranco commented:
`calcite-color` is now dogfooding components! 🐶 ⬅ _such a good boy!_

The only outstanding item now is what to do with the small scale layout ([see previous comment ☝️](https://github.com/Esri/calcite-components/pull/592#issuecomment-663354661)).

paulcpederson commented:
For now, in small scale we could just drop the `r` `g` `b` text, maybe?

driskull commented on src/components/calcite-color-hex-input/calcite-color-hex-input.tsx:
```diff
@@ -0,0 +1,231 @@
+   * Label used for the hex input.
+   */
+  @Prop()
+  hexLabel = "Hex";
```
does this need to be an `intl` named prop?

jcfranco replied:
Not sure since the text isn't displayed. For `calcite-icon` we have `textLabel`, which is used specifically for the aria attribute, so these similar props follow that pattern.

paulcpederson replied:
Since they are strings with default english backup text, I think they probably should follow our `intl-` convention. This is like the modal's close text which is used exclusively for aria-label

jcfranco replied:
Alrighty, I'll rename the props and create a `calcite-icon` issue to follow suit. 🎴

jcfranco replied:
💥

driskull commented on src/components/calcite-color/calcite-color.tsx:
```diff
@@ -0,0 +1,933 @@
+   * Label used for the delete color button.
+   */
+  @Prop()
+  deleteColorLabel = "Delete color";
```
does this need to be an `intl` named prop?

driskull commented on src/components/calcite-color/calcite-color.tsx:
```diff
@@ -0,0 +1,933 @@
+   * Label used for the save color button.
+   */
+  @Prop()
+  saveColorLabel = "Save color";
```
does this need to be an intl named prop?

bstifle commented:
ok @jcfranco 

I had to do some pretty major refactoring since those inputs werent working with the S size...

take a looksie (also provides some more detailed spacing specs)

![Color Picker Sizes](https://user-images.githubusercontent.com/14083530/88433189-79d38700-cdb2-11ea-8fee-a4d1aff23e36.png)
![Color Picker Spacing](https://user-images.githubusercontent.com/14083530/88433193-7a6c1d80-cdb2-11ea-9e3c-0d8c4bac3096.png)

bstifle commented:
ok @jcfranco here we go again...

sorry for the back and forth and universal sadness throughout this component.

But... here's the latest. Provided some spacing specs too. (all the typography is 14px) 

any questions please let me know

![Color Picker Sizes](https://user-images.githubusercontent.com/14083530/88436285-d6d23b80-cdb8-11ea-86b7-5590fb1c30da.png)
![Color Picker Spacing](https://user-images.githubusercontent.com/14083530/88436281-d5087800-cdb8-11ea-87c3-1f833ea46127.png)

bstifle commented:
@jcfranco @macandcheese @julio8a 

This is getting really confusing since we already have calcite-colors as a library.

Let's change the name back to **calcite-color-picker**

macandcheese commented:
`calcite-color` makes more sense to me... it's more consistent with `calcite-date`, and lends itself more to granular component use for things like `calcite-color-swatch` outside of the color picking context.

TheBlueDog commented:
If you are using the scale system the graph height on M and L is slightly off.

macandcheese commented on src/components/calcite-color/calcite-color.stories.js:
```diff
@@ -0,0 +1,38 @@
+  .add(
+    "Simple",
+    () => `
+    <calcite-color-picker
```
Remnant ✨

jcfranco replied:
💥
